### PR TITLE
Integrate with ppxlib

### PR DESCRIPTION
**Pull request by hhugo**
Following suggestion in https://github.com/LexiFi/gen_js_api/pull/83#issuecomment-516743718

### CODE REVIEW

**Comment by hhugo** [comment]
cc @tyoverby

**Comment by hhugo** [comment]
@nojb, this should be ready for review.

**Comment by hhugo** [comment]
gentle ping

**Comment by nojb** [comment]
@hhugo I'm away at the moment, but will take a look at this on Monday when I'm back. Thanks!

File: ppx-lib/gen_js_api_ppx.ml [hunk 1]
```diff
@@ -1509,24 +1530,28 @@ and mapper =
...
-    | Pexp_extension (attr, PTyp ty) when filter_attr_name "js.of" attr ->
+    | Pexp_extension (attr, PTyp ty) when filter_extension "js.of" attr ->
         with_default_loc e.pexp_loc
           (fun () -> ml2js_fun (parse_typ ~global_attrs ty))
     | _ ->
         e
   in
-  {super with module_expr; structure_item; expr; typ}
+  let attribute self a =
+    ignore (filter_attr "js.dummy" a : bool);
```
Comment (nojb) [inline ppx-lib/gen_js_api_ppx.ml:1546]: Can you say a few words about this change ?

Comment (hhugo) [reply ppx-lib/gen_js_api_ppx.ml:1546]: One of the reason I want to integrate with ppxlib is that ppxlib will complain about any unused attributes. In that context, using an arbitrary attribute for defining `functions that return functions` does not work. I've created a special dummy attribute for that case.

**Comment by hhugo** [comment]
gentle ping

File: ppx-lib/gen_js_api_ppx.ml [hunk 2]
```diff
@@ -1580,5 +1605,9 @@ let mapper =
     Ast_mapper.structure = (fun _this str ->
       check_loc_mapper.Ast_mapper.structure
         check_loc_mapper
-        (mapper.Ast_mapper.structure mapper str))
+        (mapper.Ast_mapper.structure mapper str));
+    signature = (fun this s ->
+      (* run [str_of_sg] to mark `js.***` attributes as used. *)
+      let (_ : structure) = str_of_sg ~global_attrs:[] s in
```
Comment (nojb) [inline ppx-lib/gen_js_api_ppx.ml]: This call to `str_of_sg` breaks the `ppx` mode of `gen_js_api`: for example, every sum type needs to be annotated with a `js.*` attribute, etc. In fact, the `ppx` mode is not supposed to do anything on `.mli` files, see https://github.com/LexiFi/gen_js_api/blob/master/PPX.md for the details.

Perhaps this check can be moved to the `ppxlib` driver ?

**Review by nojb — CHANGES_REQUESTED** [review]
Sorry for the horrible lag, but I finally got around to reviewing this. There is a small non-backwards compatible change in the behaviour of `-ppx` mode with respect to the treatment of `.mli` files.

Other than that, seems OK! Thanks!

File: ppx-lib/gen_js_api_ppx.ml (hunk 2 above)
Comment (hhugo) [reply ppx-lib/gen_js_api_ppx.ml]: This is done

**Comment by hhugo** [comment]
Sorry for the long delay. I've rebased the feature and addressed your comment.
Please take a look again.

**Comment by hhugo** [comment]
The CI failure is spurious I believe.

**Review by nojb — APPROVED** [review]
LGTM

**Comment by nojb** [comment]
(We are waiting to see if the CI can be fixed)

**Comment by mlasson** [comment]
I think it was a networking issue on the travis side. Restarting the build seems to fix the issue. 
(Hint: you can close/reopen a PR to restart a build request to travis if you don't have admin rights; I restarted the build manually so no need to do that :)).

**Comment by nojb** [comment]
Thanks!

**Pull request merged**